tyrian/js/src/test/scala/tyrian/SubLawsTest.scala

```diff
@@ -0,0 +1,28 @@
+package tyrian
+
+import cats.effect.IO
+import cats.kernel.laws.discipline.EqTests
+import cats.kernel.laws.discipline.MonoidTests
+import cats.laws.discipline.FunctorTests
+import org.scalacheck.Arbitrary
+import org.scalacheck.Cogen
+import org.scalacheck.Gen
+import org.scalacheck.rng.Seed
+
+class SubLawsTests extends DisciplineSuite {
+
+  given [A: Arbitrary]: Arbitrary[Sub[IO, A]] =
+    Arbitrary(Arbitrary.arbitrary[A].map(Sub.emit))
+
+  given Cogen[Sub[IO, Int]] = Cogen {
+    case Sub.None             => 0L
+    case Sub.Observe(_, _, _) => 1L
+    case Sub.Combine(_, _)    => 2L
+    case Sub.Batch(_)         => 3L
+  }
```

Comment:
This cogen instance is suspect. You should derive it also from the contents, not just the type.

Reply:
Would something like this make sense? I could never fully understand `Cogen` 😄 

```scala
  val cogenValue: Sub[IO, Int] => Long =
    case Sub.None             => 0L
    case Sub.Observe(_, _, _) => 1L
    case Sub.Combine(x, y)    => 2L + cogenValue(x) + cogenValue(y)
    case Sub.Batch(xs)        => xs.foldLeft(3L)((acc, sub) => acc + cogenValue(sub))

  given Cogen[Sub[IO, Int]] = Cogen(cogenValue)
```

Reply:
The `Cogen` implementation should be matching your notion of equality. Usually I implement it by delegating to the cogen of other simpler structures.

Essentially can want to unpack a `Sub` like this:
```scala
Either[
  Unit, // Sub.None
  Either[
    Int, // Sub.Observe id
    Either[
      (Sub, Sub), // Sub.Combine
      List[Sub] // Sub.Batch
    ]
  ]
]
```
And then get the cogen of that.

Reply:
My brain is not ready for this challenge today 😆 I'll leave it here for now:

```scala
  type SubCogen[A] = Either[
    Unit, // Sub.None
    Either[
      String, // Sub.Observe id
      Either[
        (Sub[IO, A], Sub[IO, A]), // Sub.Combine
        List[Sub[IO, A]]          // Sub.Batch
      ]
    ]
  ]

  val cogenValue: Sub[IO, Int] => SubCogen[Int] =
    case Sub.None              => Left(())
    case Sub.Observe(id, _, _) => Right(Left(id))
    case Sub.Combine(x, y)     => Right(Right(Left(x, y)))
    case Sub.Batch(xs)         => Right(Right(Right(xs)))

  given Cogen[Sub[IO, Int]] = Cogen.cogenEither.contramap(cogenValue)
```